plugin/forward/persistent.go

```diff
@@ -96,13 +92,49 @@ Wait:
 
 			t.conns["tcp-tls"] = append(t.conns["tcp-tls"], &persistConn{conn, time.Now()})
 
+		case <-ticker.C:
+			t.cleanup(false)
+
 		case <-t.stop:
+			t.cleanup(true)
 			close(t.ret)
 			return
 		}
 	}
 }
 
+// closeConns closes connections.
+func closeConns(conns []*persistConn) {
+	for _, pc := range conns {
+		pc.c.Close()
+	}
+}
+
+// cleanup removes connections from cache.
+func (t *transport) cleanup(all bool) {
+	staleTime := time.Now().Add(-t.expire)
+	for proto, stack := range t.conns {
+		if len(stack) == 0 {
+			continue
+		}
+		if all {
+			go closeConns(stack)
+			t.conns[proto] = nil
+			continue
+		}
+		if stack[0].used.After(staleTime) {
+			continue
+		}
+
+		// connections in stack are sorted by "used"
+		good := sort.Search(len(stack), func(i int) bool {
```

Comment:
> So you're moving the complexity to this function, which would be O(n*log(n))

I believe the complexity of Search() is O(log(n)). So, this should be faster than iterating over all cached connections.

> we just want a conn that is valid, not the last one used...

yes, and the fastest way to find a valid connection is to check the last one. Since it was added last, it has higher probability (comparing to other connections) that it is not expired. If the last (in the slice) connection is expired this means all previous connections are expired as well, we even don't need to check them

Another benefit of taking connection from the end of slice is that we would potentially process higher traffic with less connections in cache (i.e. consume less system resources). Lets consider the following case.

Expiration time is 10s. We have 100 connections in cache after some previous user activity. At some point user starts sending 1 DNS query each 100ms. The DNS processing time is 10ms.

The original code will take a cached connection **from the begin** of cached conns and 10ms later it will add the connection back to the end of conns list with updated "used" timestamp. So, it will just rotate the 100 connections in the cache without releasing them, because each connection will be periodically updated with new timestamp, and no connection will expire

The new code will take a connection **from the end** of list and 10ms later the connection will be added back to the end of conns list with updated "used" timestamp. On next request we will use same connection again. After 10s all connections except the last one will get expired and released (by timer). Only one connection will reside in cache and this single connection would be enough to process current load (i.e. 1 query per 100ms)

So, I would say the new approach is more adaptive to the load and more prudent about system resources

Reply:
ok, trust you on this.